onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+    ind.tree_id = static_cast<int>(target_class_treeids[i]);
+    ind.node_id = static_cast<int>(target_class_nodeids[i]);
+    if (idi.find(ind) == idi.end()) {
+      ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (weights).");
+    }
+    w.i = target_class_ids[i];
+    w.value = target_class_weights[i];
+    idi[ind]->weights.push_back(w);
+  }
+
+  n_trees_ = roots_.size();
+  has_missing_tracks_ = false;
+  for (auto itm = nodes_missing_value_tracks_true.begin();
+       itm != nodes_missing_value_tracks_true.end(); ++itm) {
+    if (*itm) {
+      has_missing_tracks_ = true;
+      break;
+    }
+  }
+}
+
+template <typename ITYPE, typename OTYPE>
+void TreeEnsembleCommon<ITYPE, OTYPE>::compute(const Tensor* X, Tensor* Z, Tensor* label) const {
+  switch (aggregate_function_) {
+    case AGGREGATE_FUNCTION::AVERAGE:
+      compute_agg(
+          X, Z, label,
+          _AggregatorAverage<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+    case AGGREGATE_FUNCTION::SUM:
+      compute_agg(
+          X, Z, label,
+          _AggregatorSum<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+    case AGGREGATE_FUNCTION::MIN:
+      compute_agg(
+          X, Z, label,
+          _AggregatorMin<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+    case AGGREGATE_FUNCTION::MAX:
+      compute_agg(
+          X, Z, label,
+          _AggregatorMax<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+  }
+  ORT_THROW("Unknown aggregation function in TreeEnsemble.");
+}
+
+template <typename ITYPE, typename OTYPE>
+template <typename AGG>
+void TreeEnsembleCommon<ITYPE, OTYPE>::compute_agg(const Tensor* X, Tensor* Z, Tensor* label, const AGG& agg) const {
+  int64_t stride = X->Shape().NumDimensions() == 1 ? X->Shape()[0] : X->Shape()[1];
+  int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
+
+  const ITYPE* x_data = X->template Data<ITYPE>();
+  OTYPE* z_data = Z->template MutableData<OTYPE>();
+  int64_t* label_data = label == NULL ? NULL : label->template MutableData<int64_t>();
+
+  if (n_targets_or_classes_ == 1) {
+    if (N == 1) {
+      OTYPE scores = 0;
+      unsigned char has_scores = 0;
+      if (n_trees_ <= omp_tree_) {
+        for (int64_t j = 0; j < n_trees_; ++j)
+          agg.ProcessTreeNodePrediction1(
+              &scores,
+              ProcessTreeNodeLeave(roots_[j], x_data),
+              &has_scores);
+      } else {
+        std::vector<OTYPE> scores_t(n_trees_, 0);
+        std::vector<unsigned char> has_scores_t(n_trees_, 0);
+#ifdef USE_OPENMP
+#pragma omp parallel for
+#endif
+        for (int64_t j = 0; j < n_trees_; ++j) {
```

Comment:
Please use Threadpool::[Batch]ParallelFor. You can get the Threadpool from OpKernelContext and pass it in here.

Reply:
As discussed will be handled in a separate PR.